dao/src/main/scala/za/co/absa/enceladus/dao/MenasException.scala

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2018 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.enceladus.dao
+
+abstract class MenasException(message: String, cause: Throwable) extends Exception(message, cause)
+
+abstract class RetryableException(message: String, cause: Throwable) extends MenasException(message, cause)
+
+abstract class OptionallyRetryableException(message: String, cause: Throwable) extends MenasException(message, cause)
+
+object RetryableException {
+
+  type RetryableExceptionsType = Class[_ <: MenasException]
```

Comment:
Type removed, see my next commit. 

--
Regarding adding Set: I wouldn't be able to use it in `mapIntToOptionallyRetryableException` if it's a set.

--
Why `MenasException` and not `RetryableException` - I wasn't able to make the `union` (in the old version of `CrossHostApiCaller.scala` in this PR) work with `RetryablException` because of the type system - but it was fixed by using `++` and not `union`. I think that the reason is a type broadening trick that's used in `++` method:

`def ++[B >: A, That](that: GenTraversableOnce[B])(implicit bf: CanBuildFrom[Repr, B, That]): That = {`

(jfyi definition of the traversable collection that's used is: `trait TraversableLike[+A, +Repr]`)